components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts

```diff
@@ -0,0 +1,335 @@
+import {
+  Component,
+  EventEmitter,
+  Input,
+  OnDestroy,
+  OnInit
+} from '@angular/core';
+import { Subject, combineLatest } from 'rxjs';
+import { Store } from '@ngrx/store';
+import { FormBuilder,  Validators, FormGroup } from '@angular/forms';
+import {  takeUntil, filter } from 'rxjs/operators';
+import { NgrxStateAtom } from 'app/ngrx.reducers';
+import { Regex } from 'app/helpers/auth/regex';
+import {
+    saveStatus,
+    saveError
+} from 'app/entities/environments/environment.selectors';
+import { isNil } from 'lodash/fp';
+import { EntityStatus } from 'app/entities/entities';
+import { HttpStatus } from 'app/types/types';
+import { Environment } from 'app/entities/environments/environment.model';
+import { TelemetryService } from 'app/services/telemetry/telemetry.service';
+import { GetCookbooks } from 'app/entities/cookbooks/cookbook.actions';
+import { Cookbook } from 'app/entities/cookbooks/cookbook.model';
+import {
+  allCookbooks,
+  getAllStatus as getAllCookbooksForOrgStatus
+} from 'app/entities/cookbooks/cookbook.selectors';
+import { CreateEnvironment, GetEnvironments } from 'app/entities/environments/environment.action';
+
+const CREATE_TAB_NAME = 'environmentTab';
+
+export class DynamicGrid {
+  id: number;
+  name: string;
+  operator: string;
+  version: string;
+}
+
+@Component({
+  selector: 'app-create-environment-modal',
+  templateUrl: './create-environment-modal.component.html',
+  styleUrls: ['./create-environment-modal.component.scss']
+})
+
+export class CreateEnvironmentModalComponent implements OnInit, OnDestroy {
+
+  @Input() openEvent: EventEmitter<void>;
+  @Input() environmentsList: Environment[] = [];
+  @Input() serverId: string;
+  @Input() orgId: string;
+
+  public visible = false;
+  public creating = false;
+  public conflictError = false;
+  public constraintsTab = false;
+  public detailsTab = true;
+  public isLinear = true;
+  public defaultTab = false;
+  public overrideTab = false;
+  public showdrag = false;
+  public showConstraint = false;
+
+  public detailsFormGroup: FormGroup;
+  public defaultAttrFormGroup: FormGroup;
+  public overrideAttrFormGroup: FormGroup;
+  public constraintKeys: string[] = [];
+  public cookbooks: Cookbook[] = [];
+  public attr_value = '{}';
+  public server: string;
+  public org: string;
+  public per_page = 9;
+  public page = 1;
+  public name_id = '';
+  public jsonString: string;
+  public defaultAttrParseError: boolean;
+  public overrideAttrParseError: boolean;
+  public data: any;
```

Comment:
Can we use a strong type here?

Reply:
i have updated the onjsonchange function and no need to use this variable.